src/System.Runtime.Numerics/src/System/Numerics/BigCalc.SquMul.cs

```diff
@@ -0,0 +1,356 @@
+
+            var bits = new uint[valueLength * 2];
+
+            fixed (uint* v = value, b = bits)
+            {
+                Square(v, valueLength,
+                       b, bits.Length);
+            }
+
+            return bits;
+        }
+
+        [SecuritySafeCritical]
+        private unsafe static void Square(uint* value, int valueLength,
+                                          uint* bits, int bitsLength)
+        {
+            Debug.Assert(valueLength >= 0);
+            Debug.Assert(bitsLength == valueLength * 2);
+
+            if (valueLength < SquareThreshold)
+            {
+                // squares the bits
+                for (var i = 0; i < valueLength; i++)
+                {
+                    var carry = 0UL;
+                    for (var j = 0; j < i; j++)
+                    {
+                        var digit1 = bits[i + j] + carry;
+                        var digit2 = (ulong)value[j] * value[i];
+                        bits[i + j] = (uint)(digit1 + (digit2 << 1));
+                        carry = (digit2 + (digit1 >> 1)) >> 31;
+                    }
+                    var digits = (ulong)value[i] * value[i] + carry;
+                    bits[i * 2] = (uint)digits;
+                    bits[i * 2 + 1] = (uint)(digits >> 32);
+                }
+            }
+            else
+            {
+                // divide & conquer
+                // (see multiply...)
+                var n = valueLength / 2;
+
+                var valueLow = value;
+                var valueLowLength = n;
+                var valueHigh = value + n;
+                var valueHighLength = valueLength - n;
+
+                var bitsLow = bits;
+                var bitsLowLength = n * 2;
+                var bitsHigh = bits + n * 2;
+                var bitsHighLength = bitsLength - n * 2;
+
+                Square(valueLow, valueLowLength,
+                       bitsLow, bitsLowLength);
+                Square(valueHigh, valueHighLength,
+                       bitsHigh, bitsHighLength);
+
+                var fold = AddFold(valueHigh, valueHighLength,
+                                   valueLow, valueLowLength);
+                var foldLength = fold[fold.Length - 1] == 0
+                               ? fold.Length - 1 : fold.Length;
+
+                fixed (uint* core = Square(fold, foldLength))
+                {
+                    var coreLength = foldLength * 2;
+                    if (core[coreLength - 1] == 0)
+                        coreLength = coreLength - 1;
+
+                    SubtractCore(bitsHigh, bitsHighLength,
+                                 bitsLow, bitsLowLength,
+                                 core, coreLength);
+
+                    // merge the result
+                    AddSelf(bits + n, bitsLength - n, core, coreLength);
+                }
+            }
+        }
+
+        public static uint[] Multiply(uint left, uint right)
+        {
+            var bits = new uint[2];
```

Comment:
Would it make sense to return two uint out parameters instead of allocating the array?